platooning_strategic_IHP/src/platoon_strategic_ihp.cpp

```diff
@@ -2501,52 +2531,75 @@ namespace platoon_strategic_ihp
                 pm_.current_platoon_state = PlatoonState::LEADER;
-                ROS_DEBUG_STREAM("Changed the state back to Leader");
 
                 // End the method if time out.
+                pm_.clearActionPlan();
+                pm_.currentPlatoonID = "";
+                pm_.platoonLeaderID = "";
+                pm_.targetPlatoonID = "";
                 return;
             } 
         }
 
         // Task 3: update plan: PLATOON_FRONT_JOIN with new gap
         double desiredJoinGap2 = config_.desiredJoinTimeGap * current_speed_;
         double maxJoinGap = std::max(config_.desiredJoinGap, desiredJoinGap2);
+
         // check if compatible for front join --> return front gap, no veh type check, is compatible
-        double currentGap = pm_.getDistanceToPredVehicle();  //TODO: fix this! it always returns 0.
+        // Note that pm_ only represents host's platoon members. As the aborting leader, the only vehicle
+        // preceding host is the candidate joiner. For this code to work, it depends on the candidate to publish
+        // mobility operation STATUS messages so that host can include it in the pm_ platoon membership.
+        double currentGap = pm_.getDistanceToPredVehicle();
         ROS_DEBUG_STREAM("Based on desired join time gap, the desired join distance gap is " << desiredJoinGap2 << " m");
         ROS_DEBUG_STREAM("Since we have max allowed gap as " << config_.desiredJoinGap << " m then max join gap became " << maxJoinGap << " m");
         ROS_DEBUG_STREAM("The current gap to joiner is " << currentGap << " m");
         ROS_DEBUG_STREAM("current_plan valid = " << pm_.current_plan.valid << ", isFirstLeaderAbortRequest = " << isFirstLeaderAbortRequest_);
 
-        // Check if gap is big enough and if current plan is timeout.
+        // NOTE: The front join depends upon the joiner to publish op STATUS messages with this platoon's ID, then host receives at least one
+        // and thereby adds the joiner to the platoon record. This process requires host's mob_req_cb_leader() to ACK the join request, then 
+        // the remote vehicle to handle that ACK and broadcast its first op STATUS, then host to receive and process it. All that has to happen
+        // before the below code block runs, even though this method starts to spin immediately after we send out the mentioned ACK. To avoid
+        // a race condition, we must wait a few cycles to ensure the 2-way messaging has completed.
+
+        // Check if gap is big enough and if there is no currently active plan and this method has been called several times
         // Add a condition to prevent sending repeated requests (Note: This is a same-lane maneuver, so no need to consider lower bound of joining gap.)
-        if (currentGap <= maxJoinGap  &&  !pm_.current_plan.valid  &&  isFirstLeaderAbortRequest_) 
+        ++numLeaderAbortingCalls_;
+        if (currentGap <= maxJoinGap  &&  !pm_.current_plan.valid  &&  numLeaderAbortingCalls_ > 4) 
```

Comment:
I dont understand how the request being sent late is going to prevent a race condition? since the request is the first step of frontal join

Reply:
This is the second request in the frontal join process (near the bottom of the sequence diagram).  After the first request/ACK completes, then the host (former lead vehicle) must see at least one STATUS message from the joiner before forming this request message.  If that does not happen, the calculation of currentGap on line 2564 returns 0 since it doesn't see the joiner yet.  That result would allow the second request to go out prematurely and complete the join even if the vehicle isn't in position.  I've expanded the NOTE comment above to help clarify the situation.  Let me know if better/more wording is needed.

Reply:
now that you explain it, the comment above is clear. Another question, from what I see, unless the speed the joinGap on line 2564 is never zero?
Also, I am concerned that by the time numLeaderAbortingCalls_ is higher than 4, the leaderaborting state is timed out! so maybe we need to make that 4 a configurable param?

Reply:
Not sure what your first question is.  I don't see maxJoinGap ever being zero.  My original explanation referred to the wrong line, which should have been the calculation of currentGap, now on line 2570.  The concern is that currentGap will be zero unless we've seen a STATUS message from the joiner.
Agree, the call threshold should be a param.  I'll make that change.